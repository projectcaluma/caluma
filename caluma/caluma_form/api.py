from typing import Any, Optional

from caluma.caluma_form import domain_logic, models
from caluma.caluma_user.models import BaseUser


def save_answer(
    question: models.Question,
    document: Optional[models.Document] = None,
    user: Optional[BaseUser] = None,
    value: Optional[Any] = None,
    context: Optional[dict] = None,
    **kwargs
) -> models.Answer:
    """
    Save an answer for given question, document.

    Similar to saveDocumentStringAnswer and the likes, it performes upsert.
    :param value: Must match the question type
    """

    data = {"question": question, "document": document, "value": value}
    data.update(kwargs)

    answer = models.Answer.objects.filter(question=question, document=document).first()
    answer = domain_logic.SaveAnswerLogic.get_new_answer(data, user, answer)

    return answer


def save_default_answer(
    question: models.Question,
    user: Optional[BaseUser] = None,
    value: Optional[Any] = None,
    **kwargs
) -> models.Answer:
    """
    Save default_answer for given question.

    Similar to saveDefaultStringAnswer and the likes, it performes upsert.
    :param value: Must match the question type
    """

    data = {"question": question, "value": value}
    data.update(kwargs)

    answer = question.default_answer
    answer = domain_logic.SaveDefaultAnswerLogic.get_new_answer(data, user, answer)

    return answer


def save_document(
    form: models.Form,
    meta: Optional[dict] = None,
    document: Optional[models.Document] = None,
    user: Optional[BaseUser] = None,
) -> models.Document:
    """Save a document for a given form."""

    if meta is None:
        meta = {}

    if not document:
        return domain_logic.SaveDocumentLogic.create(
            {"form": form, "meta": meta}, user=user
        )

    domain_logic.SaveDocumentLogic.update(
        document, {"form": form, "meta": meta}, user=user
    )
    return document


def copy_form(
    source: models.Form,
    slug: str,
    name: Optional[str] = None,
    description: Optional[str] = None,
    is_published: Optional[bool] = None,
    user: Optional[BaseUser] = None,
) -> models.Form:
    """Copy a form."""

    return domain_logic.CopyFormLogic.copy(
        {
            "source": source,
            "slug": slug,
            "name": name,
            "description": description,
            "is_published": False if is_published is None else is_published,
        },
        user=user,
    )
