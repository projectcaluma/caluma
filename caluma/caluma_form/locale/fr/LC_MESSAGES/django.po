# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-09-05 12:21+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n > 1);\n"

#: caluma/caluma_form/format_validators.py:58
msgid "E-mail address"
msgstr "Adresse e-mail"

#: caluma/caluma_form/format_validators.py:60
msgid "Please enter a valid e-mail address"
msgstr "Veuillez entrer une addresse e-mail valide"

#: caluma/caluma_form/format_validators.py:65
msgid "Phone number"
msgstr "Numéro de téléphone"

#: caluma/caluma_form/format_validators.py:67
msgid "Please enter a valid phone number"
msgstr "Veuillez entrer un numéro de téléphone valide"
