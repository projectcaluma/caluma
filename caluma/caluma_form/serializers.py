from django.db import transaction
from rest_framework import exceptions
from rest_framework.serializers import (
    CharField,
    DateField,
    FloatField,
    IntegerField,
    ListField,
    PrimaryKeyRelatedField,
)

from ..caluma_core import serializers
from . import domain_logic, models, validators
from .jexl import QuestionJexl


class QuestionJexlField(serializers.JexlField):
    def __init__(self, **kwargs):
        super().__init__(QuestionJexl(), **kwargs)


class SaveFormSerializer(serializers.ModelSerializer):
    class Meta:
        model = models.Form
        fields = ["slug", "name", "description", "meta", "is_archived", "is_published"]


class CopyFormSerializer(serializers.ModelSerializer):
    source = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Form.objects, required=True
    )

    @transaction.atomic
    def create(self, validated_data):
        return domain_logic.CopyFormLogic.copy(
            validated_data, user=self.context["request"].user
        )

    class Meta:
        model = models.Form
        fields = ["slug", "name", "description", "source", "is_published"]


class AddFormQuestionSerializer(serializers.ModelSerializer):
    form = serializers.GlobalIDField(source="slug")
    question = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Question.objects
    )

    def update(self, instance, validated_data):
        # default sort is 0, as per default form question are sorted
        # in descending order this will be at the end
        _, created = models.FormQuestion.objects.get_or_create(
            form=self.instance, question=validated_data["question"]
        )

        if created:
            # reassign sort from start 1 so a newly added item with sort 0 will
            # be at the end again
            for sort, question in enumerate(
                self.instance.questions.all().order_by("formquestion__sort"), start=1
            ):
                models.FormQuestion.objects.filter(
                    form=instance, question=question
                ).update(sort=sort)

        return instance

    class Meta:
        fields = ["form", "question"]
        model = models.Form


class RemoveFormQuestionSerializer(serializers.ModelSerializer):
    form = serializers.GlobalIDField(source="slug")
    question = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Question.objects
    )

    def update(self, instance, validated_data):
        models.FormQuestion.objects.filter(
            form=instance, question=validated_data["question"]
        ).delete()
        return instance

    class Meta:
        fields = ["form", "question"]
        model = models.Form


class FormQuestionRelatedField(serializers.GlobalIDPrimaryKeyRelatedField):
    def get_queryset(self):
        form = self.parent.parent.instance
        return form.questions.all()


class ReorderFormQuestionsSerializer(serializers.ModelSerializer):
    form = serializers.GlobalIDField(source="slug")
    questions = FormQuestionRelatedField(many=True)

    def update(self, instance, validated_data):
        questions = validated_data["questions"]
        curr_questions = set(instance.questions.all())

        if len(questions) != len(curr_questions) or set(questions) - set(
            instance.questions.all()
        ):
            raise exceptions.ValidationError(
                "Input questions are not the same as current form questions"
            )

        for sort, question in enumerate(reversed(questions), start=1):
            models.FormQuestion.objects.filter(form=instance, question=question).update(
                sort=sort
            )

        return instance

    class Meta:
        fields = ["form", "questions"]
        model = models.Form


class CopyQuestionSerializer(serializers.ModelSerializer):
    source = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Question.objects, required=True
    )

    @transaction.atomic
    def create(self, validated_data):
        user = self.context["request"].user
        source = validated_data["source"]
        validated_data["type"] = source.type
        validated_data["is_required"] = source.is_required
        validated_data["is_hidden"] = source.is_hidden
        validated_data["configuration"] = dict(source.configuration)
        validated_data["meta"] = dict(source.meta)
        validated_data["row_form"] = source.row_form

        question = super().create(validated_data)

        new_question_options = [
            models.QuestionOption(
                sort=sort,
                question=question,
                option=question_option.option,
                created_by_user=user.username,
                created_by_group=user.group,
                modified_by_user=user.username,
                modified_by_group=user.group,
            )
            for sort, question_option in enumerate(
                reversed(models.QuestionOption.objects.filter(question=source)), start=1
            )
        ]
        for question_option in new_question_options:
            question_option.save()

        return question

    class Meta:
        model = models.Question
        fields = ["slug", "label", "source"]


class SaveQuestionSerializer(serializers.ModelSerializer):
    is_hidden = QuestionJexlField(required=False)
    is_required = QuestionJexlField(required=False)

    def validate(self, data):
        validators.QuestionValidator().validate(data)
        return super().validate(data)

    class Meta:
        model = models.Question
        fields = [
            "slug",
            "label",
            "info_text",
            "is_required",
            "is_hidden",
            "meta",
            "is_archived",
        ]


class SaveTextQuestionSerializer(SaveQuestionSerializer):
    min_length = IntegerField(min_value=1, required=False, allow_null=True)
    max_length = IntegerField(min_value=1, required=False, allow_null=True)
    format_validators = ListField(child=CharField(), required=False)

    def validate(self, data):
        data["type"] = models.Question.TYPE_TEXT
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + [
            "min_length",
            "max_length",
            "placeholder",
            "format_validators",
        ]


class SaveTextareaQuestionSerializer(SaveQuestionSerializer):
    min_length = IntegerField(min_value=1, required=False, allow_null=True)
    max_length = IntegerField(min_value=1, required=False, allow_null=True)
    format_validators = ListField(child=CharField(), required=False)

    def validate(self, data):
        data["type"] = models.Question.TYPE_TEXTAREA
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + [
            "min_length",
            "max_length",
            "placeholder",
            "format_validators",
        ]


class SaveDateQuestionSerializer(SaveQuestionSerializer):
    def validate(self, data):
        data["type"] = models.Question.TYPE_DATE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields


class SaveQuestionOptionsMixin(object):
    def create_question_options(self, question, options):
        user = self.context["request"].user
        question_options = [
            models.QuestionOption(
                sort=sort,
                question=question,
                option=option,
                created_by_user=user.username,
                created_by_group=user.group,
                modified_by_user=user.username,
                modified_by_group=user.group,
            )
            for sort, option in enumerate(reversed(options), start=1)
        ]
        for question_option in question_options:
            question_option.save()

    @transaction.atomic
    def create(self, validated_data):
        options = validated_data.pop("options")
        instance = super().create(validated_data)
        self.create_question_options(instance, options)
        return instance

    @transaction.atomic
    def update(self, instance, validated_data):
        options = validated_data.pop("options")
        models.QuestionOption.objects.filter(question=instance).delete()
        instance = super().update(instance, validated_data)
        self.create_question_options(instance, options)
        return instance


class SaveMultipleChoiceQuestionSerializer(
    SaveQuestionOptionsMixin, SaveQuestionSerializer
):
    options = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Option.objects.all(), many=True, required=True
    )

    def validate(self, data):
        data["type"] = models.Question.TYPE_MULTIPLE_CHOICE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["options"]


class SaveChoiceQuestionSerializer(SaveQuestionOptionsMixin, SaveQuestionSerializer):
    options = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Option.objects.all(), many=True, required=True
    )

    def validate(self, data):
        data["type"] = models.Question.TYPE_CHOICE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["options"]


class SaveDynamicChoiceQuestionSerializer(SaveQuestionSerializer):
    data_source = CharField()

    def validate(self, data):
        data["type"] = models.Question.TYPE_DYNAMIC_CHOICE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["data_source"]


class SaveDynamicMultipleChoiceQuestionSerializer(SaveQuestionSerializer):
    data_source = CharField()

    def validate(self, data):
        data["type"] = models.Question.TYPE_DYNAMIC_MULTIPLE_CHOICE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["data_source"]


class SaveFloatQuestionSerializer(SaveQuestionSerializer):
    min_value = FloatField(required=False, allow_null=True)
    max_value = FloatField(required=False, allow_null=True)

    def validate(self, data):
        min_value = (
            data.get("min_value")
            if data.get("min_value") is not None
            else float("-inf")
        )
        max_value = (
            data.get("max_value") if data.get("max_value") is not None else float("inf")
        )

        if min_value > max_value:
            raise exceptions.ValidationError(
                f"max_value {max_value} is smaller than {min_value}"
            )

        data["type"] = models.Question.TYPE_FLOAT
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + [
            "min_value",
            "max_value",
            "placeholder",
        ]


class SaveIntegerQuestionSerializer(SaveQuestionSerializer):
    min_value = IntegerField(required=False, allow_null=True)
    max_value = IntegerField(required=False, allow_null=True)

    def validate(self, data):
        min_value = (
            data.get("min_value")
            if data.get("min_value") is not None
            else float("-inf")
        )
        max_value = (
            data.get("max_value") if data.get("max_value") is not None else float("inf")
        )

        if min_value > max_value:
            raise exceptions.ValidationError(
                f"max_value {max_value} is smaller than {min_value}"
            )

        data["type"] = models.Question.TYPE_INTEGER
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + [
            "min_value",
            "max_value",
            "placeholder",
        ]


class SaveTableQuestionSerializer(SaveQuestionSerializer):
    row_form = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Form.objects,
        required=True,
        help_text=models.Question._meta.get_field("row_form").help_text,
    )

    def validate(self, data):
        data["type"] = models.Question.TYPE_TABLE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["row_form"]


class SaveFormQuestionSerializer(SaveQuestionSerializer):
    sub_form = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Form.objects, required=True
    )

    def validate(self, data):
        data["type"] = models.Question.TYPE_FORM
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + ["sub_form"]


class SaveFileQuestionSerializer(SaveQuestionSerializer):
    def validate(self, data):
        data["type"] = models.Question.TYPE_FILE
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields


class SaveStaticQuestionSerializer(SaveQuestionSerializer):
    static_content = CharField(required=False)

    def validate(self, data):
        data["type"] = models.Question.TYPE_STATIC
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = [
            "label",
            "slug",
            "info_text",
            "is_hidden",
            "meta",
            "is_archived",
            "static_content",
        ]


class SaveCalculatedFloatQuestionSerializer(SaveQuestionSerializer):
    calc_expression = QuestionJexlField(required=False)

    def validate(self, data):
        data["type"] = models.Question.TYPE_CALCULATED_FLOAT
        return super().validate(data)

    class Meta(SaveQuestionSerializer.Meta):
        fields = SaveQuestionSerializer.Meta.fields + [
            "calc_expression",
        ]


class SaveOptionSerializer(serializers.ModelSerializer):
    class Meta:
        fields = ["slug", "label", "is_archived", "meta"]
        model = models.Option


class CopyOptionSerializer(serializers.ModelSerializer):
    source = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Option.objects, required=True
    )

    def create(self, validated_data):
        source = validated_data["source"]
        validated_data["meta"] = dict(source.meta)
        return super().create(validated_data)

    class Meta:
        fields = ["slug", "label", "source"]
        model = models.Option


class DocumentSerializer(serializers.ModelSerializer):
    def create(self, validated_data):
        return domain_logic.SaveDocumentLogic.create(
            validated_data, user=self.context["request"].user
        )

    class Meta:
        model = models.Document
        fields = [
            "id",
            "form",
            "meta",
        ]


class SaveAnswerSerializer(serializers.ModelSerializer):
    def validate(self, data):
        data = domain_logic.SaveAnswerLogic.validate_for_save(
            data, self.context["request"].user, self.instance, True
        )
        return super().validate(data)

    @transaction.atomic
    def create(self, validated_data):
        return domain_logic.SaveAnswerLogic.create(
            validated_data, user=self.context["request"].user
        )

    @transaction.atomic
    def update(self, instance, validated_data):
        return domain_logic.SaveAnswerLogic.update(instance, validated_data)

    class Meta:
        model = models.Answer
        fields = ["question", "document", "meta", "value"]


class SaveDocumentStringAnswerSerializer(SaveAnswerSerializer):
    value = CharField(trim_whitespace=False, allow_blank=True, required=False)

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentListAnswerSerializer(SaveAnswerSerializer):
    value = ListField(child=CharField(), required=False)

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentIntegerAnswerSerializer(SaveAnswerSerializer):
    value = IntegerField(required=False)

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentFloatAnswerSerializer(SaveAnswerSerializer):
    value = FloatField(required=False)

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentDateAnswerSerializer(SaveAnswerSerializer):
    value = DateField(source="date", required=False)

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentTableAnswerSerializer(SaveAnswerSerializer):
    value = serializers.GlobalIDPrimaryKeyRelatedField(
        source="documents",
        queryset=models.Document.objects,
        many=True,
        required=False,
        help_text="List of document IDs representing the rows in the table.",
    )

    class Meta(SaveAnswerSerializer.Meta):
        pass


class SaveDocumentFileAnswerSerializer(SaveAnswerSerializer):
    value = CharField(write_only=True, source="file", required=False)
    value_id = PrimaryKeyRelatedField(read_only=True, source="file", required=False)

    class Meta(SaveAnswerSerializer.Meta):
        fields = SaveAnswerSerializer.Meta.fields + [
            "value_id",
        ]


class RemoveAnswerSerializer(serializers.ModelSerializer):
    answer = PrimaryKeyRelatedField(
        queryset=models.Answer.objects.filter(document__isnull=False)
    )

    @transaction.atomic
    def update(self, instance, validated_data):
        instance.delete()
        return instance

    class Meta:
        fields = [
            "answer",
        ]
        model = models.Answer


class SaveDefaultAnswerSerializer(serializers.ModelSerializer):
    def validate(self, data):
        return domain_logic.SaveDefaultAnswerLogic.validate_for_save(
            data, self.context["request"].user, self.instance, True
        )

    def create(self, validated_data):
        return domain_logic.SaveDefaultAnswerLogic.create(
            validated_data, self.context["request"].user
        )

    def update(self, instance, validated_data):
        return domain_logic.SaveDefaultAnswerLogic.update(instance, validated_data)

    class Meta:
        model = models.Answer
        fields = ["question", "meta", "value"]


class SaveDefaultStringAnswerSerializer(SaveDefaultAnswerSerializer):
    value = CharField(trim_whitespace=False, allow_blank=True, required=False)

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class SaveDefaultListAnswerSerializer(SaveDefaultAnswerSerializer):
    value = ListField(child=CharField(), required=False)

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class SaveDefaultIntegerAnswerSerializer(SaveDefaultAnswerSerializer):
    value = IntegerField(required=False)

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class SaveDefaultFloatAnswerSerializer(SaveDefaultAnswerSerializer):
    value = FloatField(required=False)

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class SaveDefaultDateAnswerSerializer(SaveDefaultAnswerSerializer):
    value = DateField(source="date", required=False)

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class SaveDefaultTableAnswerSerializer(SaveDefaultAnswerSerializer):
    value = serializers.GlobalIDPrimaryKeyRelatedField(
        source="documents",
        queryset=models.Document.objects,
        many=True,
        required=False,
        help_text="List of document IDs representing the rows in the table.",
    )

    class Meta(SaveDefaultAnswerSerializer.Meta):
        pass


class RemoveDefaultAnswerSerializer(serializers.ModelSerializer):
    question = PrimaryKeyRelatedField(queryset=models.Question.objects)

    @transaction.atomic
    def update(self, instance, validated_data):
        instance.default_answer.delete()
        return instance

    class Meta:
        fields = [
            "question",
        ]
        model = models.Question


class RemoveDocumentSerializer(serializers.ModelSerializer):
    document = serializers.GlobalIDField(source="id")

    @transaction.atomic
    def update(self, instance, validated_data):
        if hasattr(instance, "case"):
            raise Exception("You cannot remove a Document, if it's attached to a case.")

        for answer in instance.answers.filter(
            question__type=models.Question.TYPE_TABLE
        ):
            answer.documents.all().delete()

        instance.delete()

        return instance

    class Meta:
        fields = ["document"]
        model = models.Document


class CopyDocumentSerializer(serializers.ModelSerializer):
    source = serializers.GlobalIDPrimaryKeyRelatedField(
        queryset=models.Document.objects, required=True
    )

    @transaction.atomic
    def create(self, validated_data):
        return validated_data["source"].copy(
            family=None,
            user=self.context["request"].user,
        )

    class Meta:
        model = models.Document
        fields = ["source"]
