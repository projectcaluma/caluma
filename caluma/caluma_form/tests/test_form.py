import pytest
from django.utils import translation
from graphql_relay import to_global_id

from ...caluma_core.tests import extract_serializer_input_fields
from .. import models
from ..api import copy_form
from ..serializers import SaveFormSerializer


@pytest.mark.parametrize(
    "form__description,form__name,question__type",
    [("First result", "1st", models.Question.TYPE_FLOAT)],
)
def test_query_all_forms(
    db,
    snapshot,
    form,
    form_factory,
    form_question,
    form_question_factory,
    question,
    schema_executor,
):
    form_factory(name="3rd", description="Second result")
    form_factory(name="2nd", description="Second result")
    form_question_factory(form=form)

    query = """
        query AllFormsQuery($name: String, $question: String, $orderBy: [FormOrdering]) {
          allForms(name: $name, orderBy: $orderBy) {
            edges {
              node {
                id
                slug
                name
                description
                meta
                questions(search: $question) {
                  edges {
                    node {
                      id
                      slug
                      label
                    }
                  }
                }
              }
            }
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "question": question.label,
            "orderBy": ["NAME_ASC", "CREATED_AT_ASC"],
        },
    )

    assert not result.errors
    snapshot.assert_match(result.data)


@pytest.mark.parametrize("language_code", ("en", "de"))
@pytest.mark.parametrize("form__description", ("some description text", ""))
def test_save_form(db, snapshot, form, settings, schema_executor, language_code):
    query = """
        mutation SaveForm($input: SaveFormInput!) {
          saveForm(input: $input) {
            form {
              id
              slug
              name
              meta
            }
            clientMutationId
          }
        }
    """

    inp = {"input": extract_serializer_input_fields(SaveFormSerializer, form)}
    with translation.override(language_code):
        result = schema_executor(query, variable_values=inp)

    assert not result.errors
    snapshot.assert_match(result.data)


def test_save_form_created_as_admin_user(db, form, admin_schema_executor, admin_user):
    query = """
        mutation SaveForm($input: SaveFormInput!) {
          saveForm(input: $input) {
            form {
              createdByUser
            }
          }
        }
    """

    inp = {"input": extract_serializer_input_fields(SaveFormSerializer, form)}
    form.delete()  # test creation of form
    result = admin_schema_executor(query, variable_values=inp)

    assert not result.errors
    assert result.data["saveForm"]["form"]["createdByUser"] == admin_user.username


@pytest.mark.parametrize("form__meta", [{"meta": "set"}])
def test_copy_form(db, form, form_question_factory, schema_executor):
    form_question_factory.create_batch(5, form=form)
    query = """
        mutation CopyForm($input: CopyFormInput!) {
          copyForm(input: $input) {
            form {
              slug
            }
            clientMutationId
          }
        }
    """

    inp = {"input": {"source": form.pk, "slug": "new-form", "name": "Test Form"}}
    result = schema_executor(query, variable_values=inp)

    assert not result.errors

    form_slug = result.data["copyForm"]["form"]["slug"]
    assert form_slug == "new-form"
    new_form = models.Form.objects.get(pk=form_slug)
    assert new_form.name == "Test Form"
    assert new_form.meta == form.meta
    assert new_form.source == form
    assert list(
        models.FormQuestion.objects.filter(form=new_form).values("question")
    ) == list(models.FormQuestion.objects.filter(form=form).values("question"))


@pytest.mark.parametrize("form__meta", [{"meta": "set"}])
def test_copy_form_api(db, form, form_question_factory, schema_executor):
    form_question_factory.create_batch(5, form=form)

    new_form = copy_form(source=form, slug="new-form", name="Test Form")

    assert new_form.pk == "new-form"
    assert new_form.name == "Test Form"
    assert new_form.meta == form.meta
    assert new_form.source == form
    assert list(
        models.FormQuestion.objects.filter(form=new_form).values("question")
    ) == list(models.FormQuestion.objects.filter(form=form).values("question"))


def test_add_form_question(db, form, question, form_question_factory, schema_executor):
    form_questions = form_question_factory.create_batch(5, form=form)

    # initialize sorting keys
    for idx, form_question in enumerate(form_questions):
        form_question.sort = idx + 1
        form_question.save()

    query = """
        mutation AddFormQuestion($input: AddFormQuestionInput!) {
          addFormQuestion(input: $input) {
            form {
              questions {
                edges {
                  node {
                    slug
                  }
                }
              }
            }
            clientMutationId
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "input": {
                "form": to_global_id(type(form).__name__, form.pk),
                "question": to_global_id(type(question).__name__, question.pk),
            }
        },
    )

    assert not result.errors
    questions = result.data["addFormQuestion"]["form"]["questions"]["edges"]
    assert len(questions) == 6
    assert questions[-1]["node"]["slug"] == question.slug


def test_remove_form_question(
    db, form, form_question, question, snapshot, schema_executor
):
    query = """
        mutation RemoveFormQuestion($input: RemoveFormQuestionInput!) {
          removeFormQuestion(input: $input) {
            form {
              questions {
                edges {
                  node {
                    slug
                  }
                }
              }
            }
            clientMutationId
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "input": {
                "form": to_global_id(type(form).__name__, form.pk),
                "question": to_global_id(type(question).__name__, question.pk),
            }
        },
    )

    assert not result.errors
    snapshot.assert_match(result.data)


def test_reorder_form_questions(db, form, form_question_factory, schema_executor):
    form_question_factory.create_batch(2, form=form)

    query = """
        mutation ReorderFormQuestions($input: ReorderFormQuestionsInput!) {
          reorderFormQuestions(input: $input) {
            form {
              questions {
                edges {
                  node {
                    slug
                  }
                }
              }
            }
            clientMutationId
          }
        }
    """

    question_ids = (
        form.questions.order_by("slug").reverse().values_list("slug", flat=True)
    )
    result = schema_executor(
        query,
        variable_values={
            "input": {
                "form": to_global_id(type(form).__name__, form.pk),
                "questions": [
                    to_global_id(models.Question.__name__, question_id)
                    for question_id in question_ids
                ],
            }
        },
    )

    assert not result.errors
    result_questions = [
        question["node"]["slug"]
        for question in result.data["reorderFormQuestions"]["form"]["questions"][
            "edges"
        ]
    ]

    assert result_questions == list(question_ids)


def test_reorder_form_questions_invalid_question(
    db, form, question_factory, schema_executor
):

    invalid_question = question_factory()

    query = """
        mutation ReorderFormQuestions($input: ReorderFormQuestionsInput!) {
          reorderFormQuestions(input: $input) {
            form {
              questions {
                edges {
                  node {
                    slug
                  }
                }
              }
            }
            clientMutationId
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "input": {
                "form": to_global_id(type(form).__name__, form.pk),
                "questions": [
                    to_global_id(type(models.Question).__name__, invalid_question.slug)
                ],
            }
        },
    )

    assert result.errors


def test_reorder_form_questions_duplicated_question(
    db, form, question, form_question, schema_executor
):

    query = """
        mutation ReorderFormQuestions($input: ReorderFormQuestionsInput!) {
          reorderFormQuestions(input: $input) {
            form {
              questions {
                edges {
                  node {
                    slug
                  }
                }
              }
            }
            clientMutationId
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "input": {
                "form": to_global_id(type(form).__name__, form.pk),
                "questions": [question.slug, question.slug],
            }
        },
    )
    assert result.errors
