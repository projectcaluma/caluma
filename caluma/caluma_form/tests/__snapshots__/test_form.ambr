# name: test_query_all_forms[First result-1st-float]
  <class 'dict'> {
    'allForms': <class 'dict'> {
      'edges': <class 'list'> [
        <class 'dict'> {
          'node': <class 'dict'> {
            'description': 'First result',
            'id': 'Rm9ybTplbnZpcm9ubWVudGFsLXRlbg==',
            'meta': <class 'dict'> {
            },
            'name': '1st',
            'questions': <class 'dict'> {
              'edges': <class 'list'> [
                <class 'dict'> {
                  'node': <class 'dict'> {
                    'id': 'RmxvYXRRdWVzdGlvbjpyZWR1Y2UtYWdhaW5zdA==',
                    'label': 'Derek Mendoza',
                    'slug': 'reduce-against',
                  },
                },
              ],
            },
            'slug': 'environmental-ten',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'description': 'Second result',
            'id': 'Rm9ybTpsb3NzLWxlYXJu',
            'meta': <class 'dict'> {
            },
            'name': '2nd',
            'questions': <class 'dict'> {
              'edges': <class 'list'> [
              ],
            },
            'slug': 'loss-learn',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'description': 'Second result',
            'id': 'Rm9ybTpzb25nLWxpZ2h0',
            'meta': <class 'dict'> {
            },
            'name': '3rd',
            'questions': <class 'dict'> {
              'edges': <class 'list'> [
              ],
            },
            'slug': 'song-light',
          },
        },
      ],
    },
  }
---
# name: test_remove_form_question
  <class 'dict'> {
    'removeFormQuestion': <class 'dict'> {
      'clientMutationId': None,
      'form': <class 'dict'> {
        'questions': <class 'dict'> {
          'edges': <class 'list'> [
          ],
        },
      },
    },
  }
---
# name: test_save_form[-de]
  <class 'dict'> {
    'saveForm': <class 'dict'> {
      'clientMutationId': 'testid',
      'form': <class 'dict'> {
        'id': 'Rm9ybTplbnZpcm9ubWVudGFsLXRlbg==',
        'meta': <class 'dict'> {
        },
        'name': 'Bonnie Moreno',
        'slug': 'environmental-ten',
      },
    },
  }
---
# name: test_save_form[-en]
  <class 'dict'> {
    'saveForm': <class 'dict'> {
      'clientMutationId': 'testid',
      'form': <class 'dict'> {
        'id': 'Rm9ybTplbnZpcm9ubWVudGFsLXRlbg==',
        'meta': <class 'dict'> {
        },
        'name': 'Bonnie Moreno',
        'slug': 'environmental-ten',
      },
    },
  }
---
# name: test_save_form[some description text-de]
  <class 'dict'> {
    'saveForm': <class 'dict'> {
      'clientMutationId': 'testid',
      'form': <class 'dict'> {
        'id': 'Rm9ybTplbnZpcm9ubWVudGFsLXRlbg==',
        'meta': <class 'dict'> {
        },
        'name': 'Bonnie Moreno',
        'slug': 'environmental-ten',
      },
    },
  }
---
# name: test_save_form[some description text-en]
  <class 'dict'> {
    'saveForm': <class 'dict'> {
      'clientMutationId': 'testid',
      'form': <class 'dict'> {
        'id': 'Rm9ybTplbnZpcm9ubWVudGFsLXRlbg==',
        'meta': <class 'dict'> {
        },
        'name': 'Bonnie Moreno',
        'slug': 'environmental-ten',
      },
    },
  }
---
